File: matlab/examples/waveguide/Rect_WG_W_WG_Ports.m
% Rect_WG_W_WG_Ports.m
%
% Rectangular waveguide (WR-42) with generic waveguide ports using HDF5
% mode files. The TE10 field distribution is computed analytically and
% written to HDF5 files on the fly via h5writemode.
%
% This is the Matlab/Octave counterpart of the Python test
% Rect_WG_W_WG_Ports.py. Results should be indistinguishable from the
% analytic-function version in Rect_Waveguide.m.
%
% Assertions:
%   S11 < -20 dB across the passband
%   S21 > -1  dB across the passband
%
% (c) 2026 Thorsten Liebig <thorsten.liebig@gmx.de>

clc; clear; close all;

physical_constants;

% -------------------------------------------------------------------------
% Geometry  (drawing unit: 1 µm, WR-42 standard dimensions)
unit   = 1e-6;
a      = 10700;     % waveguide width  (µm)
b      = 4300;      % waveguide height (µm)
length = 50000;     % waveguide length (µm)

f_start = 20e9;
f_0     = 24e9;
f_stop  = 26e9;

lambda0  = c0 / f_0 / unit;   % free-space wavelength at f_0 (drawing units)
mesh_res = lambda0 / 30;

kc = pi / (a * unit);          % TE10 cutoff wavenumber (rad/m)

% -------------------------------------------------------------------------
% Paths
Sim_Path  = fullfile(tempdir, 'Rect_WG_HDF5');
Mode_Path = fullfile(tempdir, 'Rect_WG_HDF5_modes');
if ~exist(Mode_Path, 'dir'), mkdir(Mode_Path); end

E_file = fullfile(Mode_Path, 'TE10_E.h5');
H_file = fullfile(Mode_Path, 'TE10_H.h5');

% -------------------------------------------------------------------------
% Generate HDF5 mode files (TE10)
%   E: Vx = 0,              Vy = -sin(pi*x/a) / a
%   H: Vx = sin(pi*x/a)/a,  Vy = 0
%
% Grid runs from x=0..a and y=0..b; local_origin='corner' aligns the port
% box corner with (0,0) so mode file and simulation coordinates match.
N      = 60;
x_mode = linspace(0, a, N);
y_mode = linspace(0, b, max(round(N * b / a), 2));
nx     = numel(x_mode);
ny     = numel(y_mode);
[XX, ~] = ndgrid(x_mode, y_mode);   % XX(i,j) = x_mode(i), size (nx,ny)

Vx_E = zeros(nx, ny);
Vy_E = -sin(pi * XX / a) / a;
h5writemode(E_file, x_mode, y_mode, Vx_E, Vy_E);

Vx_H = sin(pi * XX / a) / a;
Vy_H = zeros(nx, ny);
h5writemode(H_file, x_mode, y_mode, Vx_H, Vy_H);

% -------------------------------------------------------------------------
% FDTD setup
FDTD = InitFDTD('NrTS', 1e6);
FDTD = SetGaussExcite(FDTD, 0.5*(f_start+f_stop), 0.5*(f_stop-f_start));
FDTD = SetBoundaryCond(FDTD, {'PEC','PEC','PEC','PEC','PML_8','PML_8'});

CSX = InitCSX();

mesh.x = [0, a];
mesh.y = [0, b];
mesh.z = [0, length];

% Pin mesh lines at both port faces before smoothing
z_p1_start = 10 * mesh_res;
z_p1_stop  = 15 * mesh_res;
z_p2_start = length - 10 * mesh_res;
z_p2_stop  = length - 15 * mesh_res;
mesh.z = unique([mesh.z, z_p1_start, z_p1_stop, z_p2_start, z_p2_stop]);

mesh.x = SmoothMeshLines(mesh.x, mesh_res, 1.4);
mesh.y = SmoothMeshLines(mesh.y, mesh_res, 1.4);
mesh.z = SmoothMeshLines(mesh.z, mesh_res, 1.4);

CSX = DefineRectGrid(CSX, unit, mesh);

% -------------------------------------------------------------------------
% Waveguide ports  (local_origin='corner' aligns mode file x=0,y=0 with
% the lower-left corner of the port box)
[CSX, port{1}] = AddWaveGuidePort(CSX, 0, 1, ...
    [0, 0, z_p1_start], [a, b, z_p1_stop], 'z', {}, {}, kc, 1, ...
    'E_WG_file', E_file, 'H_WG_file', H_file, 'local_origin', 'corner');

[CSX, port{2}] = AddWaveGuidePort(CSX, 0, 2, ...
    [0, 0, z_p2_start], [a, b, z_p2_stop], 'z', {}, {}, kc, 0, ...
    'E_WG_file', E_file, 'H_WG_file', H_file, 'local_origin', 'corner');

% -------------------------------------------------------------------------
% Run
CleanupSimPath(Sim_Path);

WriteOpenEMS(fullfile(Sim_Path, 'rect_wg.xml'), FDTD, CSX);
RunOpenEMS(Sim_Path, 'rect_wg.xml');

% -------------------------------------------------------------------------
% Post-processing
freq = linspace(f_start, f_stop, 201);
k    = 2*pi*freq / c0;
Z_TE = k .* Z0 ./ sqrt(k.^2 - kc^2);   % TE10 wave impedance (frequency-dependent)
port = calcPort(port, Sim_Path, freq, 'RefImpedance', Z_TE);

S11 = port{1}.uf.ref ./ port{1}.uf.inc;
S21 = port{2}.uf.ref ./ port{1}.uf.inc;

S11_dB = 20 * log10(abs(S11));
S21_dB = 20 * log10(abs(S21));

fprintf('S11 max = %.1f dB\n', max(S11_dB));
fprintf('S21 min = %.1f dB\n', min(S21_dB));

assert(max(S11_dB) < -20, sprintf('S11 too high: %.1f dB', max(S11_dB)));
assert(min(S21_dB) > -1,  sprintf('S21 too low:  %.1f dB', min(S21_dB)));

disp('PASS');

figure;
plot(freq/1e9, [S11_dB(:), S21_dB(:)], 'linewidth', 2);
grid on;
xlabel('Frequency (GHz)');
ylabel('S-Parameters (dB)');
title('Rectangular Waveguide (WR-42) with HDF5 Mode Ports');
legend('S11', 'S21');

File: python/Tests/Rect_WG_W_WG_Ports.py
"""
Rectangular waveguide (WR-42) with generic WaveguidePort using HDF5 mode files.

This is the test counterpart of Tutorials/Rect_Waveguide.py.  Instead of
AddRectWaveGuidePort (which generates the analytic mode functions internally),
it writes the identical TE10 field distributions to HDF5 mode files and uses
the generic AddWaveGuidePort.  Results should be indistinguishable.

The test is repeated for all three propagation directions (x, y, z) to verify
that the HDF5-based WaveguidePort handles arbitrary orientations correctly.
The same mode files are reused for every direction: the file coordinates
(file-x, file-y) always map to (ny_P, ny_PP) of the port.

Assertions (per direction):
  - S11 < -40 dB across the passband
  - S21 > -0.1  dB across the passband
  - mode purity > 99 % at every time step where the recorded signal exceeds
    1 % of its peak amplitude (checked for both U and I probes of both ports)

(c) 2026 Thorsten Liebig <thorsten.liebig@gmx.de>
"""

import os
import tempfile
import numpy as np
import h5py

from CSXCAD  import ContinuousStructure
from openEMS import openEMS
from openEMS.physical_constants import C0
from openEMS.utilities import check_mode_purity


def make_rect_wg_te10_hdf5(filename, field_type, a_draw, b_draw, N=60):
    """Write a TE10 rectangular waveguide mode profile to an HDF5 file.

    Parameters
    ----------
    filename : str
        Output HDF5 file path.
    field_type : 'E' or 'H'
        Which field to store.
    a_draw, b_draw : float
        Waveguide width / height in drawing units.
    N : int
        Number of grid points along the width (height scaled proportionally).
    """
    x = np.linspace(0, a_draw, N)
    y = np.linspace(0, b_draw, max(int(N * b_draw / a_draw), 2))
    XX, _ = np.meshgrid(x, y, indexing='ij')

    if field_type == 'E':
        # TE10: E_x = 0,  E_y = -sin(pi*x/a) / a
        Vx = np.zeros_like(XX)
        Vy = -np.sin(np.pi * XX / a_draw) / a_draw
    else:
        # TE10: H_x = sin(pi*x/a) / a,  H_y = 0
        Vx = np.sin(np.pi * XX / a_draw) / a_draw
        Vy = np.zeros_like(XX)

    with h5py.File(filename, 'w') as f:
        f.attrs['Version'] = 1.0
        f.create_dataset('x',  data=x)
        f.create_dataset('y',  data=y)
        f.create_dataset('Vx', data=Vx)
        f.create_dataset('Vy', data=Vy)


# ---------------------------------------------------------------------------
# Geometry & simulation parameters  (WR-42, identical to tutorial)
# ---------------------------------------------------------------------------
unit   = 1e-6       # drawing unit: 1 µm
a      = 10700      # waveguide width  (µm) — along ny_P axis
b      = 4300       # waveguide height (µm) — along ny_PP axis
length = 50000      # waveguide length (µm) — along propagation axis

f_start = 20e9
f_0     = 24e9
f_stop  = 26e9
lambda0 = C0 / f_0 / unit

mesh_res = lambda0 / 30

kc = np.pi / (a * unit)     # TE10 cutoff wavenumber (rad/m)

# ---------------------------------------------------------------------------
# Mode files — written once, reused for all directions
# (file-x maps to ny_P, file-y maps to ny_PP regardless of exc_dir)
# ---------------------------------------------------------------------------
mode_dir = os.path.join(tempfile.gettempdir(), 'Rect_WG_HDF5_modes')
os.makedirs(mode_dir, exist_ok=True)

E_file = os.path.join(mode_dir, 'TE10_E.h5')
H_file = os.path.join(mode_dir, 'TE10_H.h5')
make_rect_wg_te10_hdf5(E_file, 'E', a, b)
make_rect_wg_te10_hdf5(H_file, 'H', a, b)


def run_direction(exc_dir):
    exc_ny = 'xyz'.index(exc_dir)
    ny_P   = (exc_ny + 1) % 3
    ny_PP  = (exc_ny + 2) % 3

    Sim_Path = os.path.join(tempfile.gettempdir(), 'Rect_WG_HDF5_{}'.format(exc_dir))
    os.makedirs(Sim_Path, exist_ok=True)

    # -------------------------------------------------------------------
    # FDTD & CSX setup
    # -------------------------------------------------------------------
    FDTD = openEMS()
    FDTD.SetGaussExcite(0.5*(f_start+f_stop), 0.5*(f_stop-f_start))

    # PML on the two propagation-axis faces, PEC on the four transverse faces
    bc = [0] * 6
    bc[2*exc_ny]   = 3   # PML_8
    bc[2*exc_ny+1] = 3
    FDTD.SetBoundaryCond(bc)

    CSX = ContinuousStructure()
    FDTD.SetCSX(CSX)
    mesh = CSX.GetGrid()
    mesh.SetDeltaUnit(unit)

    mesh.AddLine('xyz'[exc_ny], [0, length])
    mesh.AddLine('xyz'[ny_P],   [0, a])
    mesh.AddLine('xyz'[ny_PP],  [0, b])

    # -------------------------------------------------------------------
    # Ports — local_origin='corner' so mode-file (x,y) aligns with
    # the (ny_P=0, ny_PP=0) corner of the port box
    # -------------------------------------------------------------------
    def make_coords(x_near, x_far):
        start = [0, 0, 0]; stop = [0, 0, 0]
        start[exc_ny] = x_near;  stop[exc_ny] = x_far
        start[ny_P]   = 0;       stop[ny_P]   = a
        start[ny_PP]  = 0;       stop[ny_PP]  = b
        return start, stop

    ports = []

    start, stop = make_coords(10*mesh_res, 15*mesh_res)
    mesh.AddLine('xyz'[exc_ny], [start[exc_ny], stop[exc_ny]])
    ports.append(FDTD.AddWaveGuidePort(1, start, stop, exc_dir,
                                        E_file=E_file, H_file=H_file,
                                        kc=kc, excite=1, local_origin='corner'))

    start, stop = make_coords(length - 10*mesh_res, length - 15*mesh_res)
    mesh.AddLine('xyz'[exc_ny], [start[exc_ny], stop[exc_ny]])
    ports.append(FDTD.AddWaveGuidePort(2, start, stop, exc_dir,
                                        E_file=E_file, H_file=H_file,
                                        kc=kc, excite=0, local_origin='corner'))

    mesh.SmoothMeshLines('all', mesh_res, ratio=1.4)

    # -------------------------------------------------------------------
    # Run
    # -------------------------------------------------------------------
    FDTD.Run(Sim_Path, cleanup=True)

    # -------------------------------------------------------------------
    # Post-processing
    # -------------------------------------------------------------------
    freq = np.linspace(f_start, f_stop, 201)
    for port in ports:
        port.CalcPort(Sim_Path, freq)

    for port in ports:
        lbl = 'port {}'.format(port.number)
        check_mode_purity(lbl + ' U', port.u_data.ui_val[0], port.u_mode_purity[0], threshold=0.99, sig_frac=0.001)
        check_mode_purity(lbl + ' I', port.i_data.ui_val[0], port.i_mode_purity[0], threshold=0.99, sig_frac=0.001)

    s11 = ports[0].uf_ref / ports[0].uf_inc
    s21 = ports[1].uf_ref / ports[0].uf_inc

    s11_dB = 20 * np.log10(np.abs(s11))
    s21_dB = 20 * np.log10(np.abs(s21))

    print('  S11 max = {:.1f} dB'.format(np.max(s11_dB)))
    print('  S21 min = {:.1f} dB'.format(np.min(s21_dB)))

    assert np.max(s11_dB) < -40, \
        'FAIL [{}]: S11 too high: {:.1f} dB'.format(exc_dir, np.max(s11_dB))
    assert np.min(s21_dB) > -0.1, \
        'FAIL [{}]: S21 too low: {:.1f} dB'.format(exc_dir, np.min(s21_dB))


for direction in ('x', 'y', 'z'):
    print('Testing direction: {}'.format(direction))
    run_direction(direction)
    print('PASS [{}]'.format(direction))
